presto-main/src/main/java/com/facebook/presto/execution/executor/TaskExecutor.java

```diff
@@ -642,6 +642,17 @@ public void run()
                         }
                         splitFinished(split);
                     }
+                    finally {
+                        // Clear the interrupted flag on the current thread, driver cancellation may have triggered an interrupt
+                        // without TaskExecutor being shut down
+                        if (Thread.interrupted()) {
+                            if (closed) {
+                                // reset interrupted flag if TaskExecutor was closed, since that interrupt may have been the
+                                // shutdown signal to this TaskRunner thread
+                                Thread.currentThread().interrupt();
+                            }
+                        }
+                    }
```

Comment:
logically this makes sense, although wondering why we can't catch the InterruptedException and set the interrupt flag if `closed` inside the catch block ?

-------------

Taking a step back, I wanted to understand the task cancellation in a little more depth.
`TaskRunners` are added via `addRunnerThread`.
https://github.com/prestodb/presto/blob/e05a499861d7747420077d456d6fc6aebf432f72/presto-main/src/main/java/com/facebook/presto/execution/executor/TaskExecutor.java#L333-L340

Within `TaskRunner` there is the standard check for interrupted, catch Interrupt and set the flag sequence.
https://github.com/prestodb/presto/blob/e05a499861d7747420077d456d6fc6aebf432f72/presto-main/src/main/java/com/facebook/presto/execution/executor/TaskExecutor.java#L584-L592

This is where I'm a bit confused. Where does the finally block you have added fit in.?
 There is a catch all Throwable block and then your finally.
 https://github.com/prestodb/presto/blob/e05a499861d7747420077d456d6fc6aebf432f72/presto-main/src/main/java/com/facebook/presto/execution/executor/TaskExecutor.java#L632-L656
 
Is this to handle the split processing which can get interrupted and you want to handle that as a separate case from the outer `TaskRunner` thread getting interrupted.?

---------
> Instead of allowing the interrupt to terminate the current task runner's
loop and re-creating a new runner to replace the interrupted one, we can
clear the interrupt flag after each iteration. Otherwise, TaskRunners
that were interrupted would have to replace themselves and end up
creating unnecessary threads in the cached threadpool executor in the
process.

What is the current behavior?. For every interrupted thread, do we end up creating a new thread in the cached thread pool.? How common is this and what is the impact.

Reply:
The answers to those questions are a little bit interrelated, so I'll give point responses and try to tie them back together at the end.

> logically this commit makes sense, although wondering why we can't catch the InterruptedException and set the interrupt flag if `closed` inside the catch block ?

There's no guarantee than an `InterruptedException` will have been thrown, in fact- it's almost certain that if an interrupt occurred to cancel the split processing, that something will have re-set the interrupted flag and re-thrown some other exception instead. 


> Is this to handle the split processing which can get interrupted and you want to handle that as a separate case from the outer TaskRunner thread getting interrupted.?

That is correct. There are exactly two sources of interrupts that could arrive here:
1. `TaskRunner` was interrupted because it was running a driver that had `Driver#close()` called on it, so the caller could not immediately destroyed the driver itself (eg: task cancellation). When this occurs and `TaskExecutor` has not itself been closed, the "signal" was just to stop a single split but not to terminate all split processing
2. `TaskExecutor` is shutdown and all threads are interrupted to terminate all `TaskRunner` activity. When this happens, `TaskExecutor#closed` will be set to true _before_ the interrupt so we can distinguish this case from scenario 1 by noting that the `TaskRunner` was not interrupted as part of `waitingSplits.take()`, otherwise that interrupt would have been handled and returned as part of that call. So the interrupt must have occurred after getting a split from the queue.

> What is the current behavior?. For every interrupted thread, do we end up creating a new thread in the cached thread pool.? How common is this and what is the impact.

The current behavior is that `TaskRunner` instances replace themselves by calling `addRunnerThread` in their `finally` block, but do so on top of a cached threadpool executor- so all calls to `executor.execute` will either transfer execution of the `Runnable` directly to an idle waiting thread or create a new thread immediately to process the submission. The thread calling `executor.execute` will of course not be able to be both the calling and receiving thread- so the new runnable will transfer to some other thread at that point and one will be created if necessary.

Now, the above isn't _terrible_ because the previously active worker threads are kept around as idle threads to potentially take another `Runnable` and satisfy future `executor.execute` calls if one arrives within the idle timeout period of 60 seconds, but those extra native threads are still provably unnecessary to create in the first place since we could have just kept the existing `TaskRunner` instance going instead.

This is... extremely common. Any time a task is cancelled because, eg: a query contains a `LIMIT` clause- all running splits will kill their active `TaskRunner` today. That said, there's an upper bound because the worst case is that all `TaskRunner` threads are interrupted while running (no more splits can be running to interrupt than there are task runner threads) and all trigger new thread creations simultaneously before any are returned to the executor as an idle worker available for reuse- so ~2x the configured worker thread creations in the worst case.

Reply:
Thanks @pettyjamesm for the detailed explanation.
One recommendation, is to add a test result section in this PR with and without the fix .
The test result section would contain the number of threads in the threadpool from the jstack output. That way we can quantify the impact of this change.

Reply:
One wrinkle with that: the same cached thread pool executor is also used whenever any splits are unblocked to trigger the blocked callback- so there’s still going to be some number of additional threads, just not as many as there might otherwise have been.